diffusion/models/text_encoder.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2022 MosaicML Diffusion authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Text encoders and tokenizers used in diffusion models."""
+
+import textwrap
+from typing import List, Optional, Tuple, Union
+
+import torch
+from transformers import AutoModel, AutoTokenizer, CLIPTextModel, CLIPTextModelWithProjection, PretrainedConfig
+
+
+class MultiTextEncoder(torch.nn.Module):
+    """Wrapper to handle multiple HuggingFace text encoders.
+
+    Wraps any number of HuggingFace text encoders to behave as one model by sharing inputs and concatenating output.
+
+    Args:
+        model_names (str, Tuple[str, ...]): Name(s) of the text encoder(s) to load. The name format should be
+            "org_name/repo_name/subfolder" where the subfolder is excluded if it is not used in the repo.
+        model_dim_keys (optional, str, list[str]): Key(s) that specify the models' output dimension in the config.
+            If ``None``, this is set to ['projection_dim', 'd_model', 'hidden_size']. Default: ``None``.
+        encode_latents_in_fp16 (bool): Whether to encode text embeddings in fp16. Default: ``True``.
+        pretrained_sdxl (bool): Whether this text encoder is for a pretrained SDXL. If true, this will only use
+            the projected output from a CLIPTextModelWithProjection. Default: ``False``.
+    """
+
+    def __init__(
+        self,
+        model_names: Union[str, Tuple[str, ...]],
+        model_dim_keys: Optional[Union[str, List[str]]] = None,
+        encode_latents_in_fp16: bool = True,
+        pretrained_sdxl: bool = False,
+    ):
+        super().__init__()
+        self.pretrained_sdxl = pretrained_sdxl
+
+        if isinstance(model_names, str):
+            model_names = (model_names,)
+        if model_dim_keys is None:
+            model_dim_keys = ['projection_dim', 'd_model', 'hidden_size']  # CLIP, T5, E5
+        torch_dtype = torch.float16 if encode_latents_in_fp16 else None
+
+        self.text_encoders = torch.nn.ModuleList()
+        self.text_encoder_dim = 0
+        self.architectures = []
+        for model_name in model_names:
+            # Process model_name string and get model config
+            name_split = model_name.split('/')
+            base_name = '/'.join(name_split[:2])
+            subfolder = '/'.join(name_split[2:])
+            text_encoder_config = PretrainedConfig.get_config_dict(base_name, subfolder=subfolder)[0]
+
+            # Add text_encoder output dim to total dim
+            dim_found = False
+            for model_dim_key in model_dim_keys:
+                if model_dim_key in text_encoder_config:
+                    self.text_encoder_dim += text_encoder_config[model_dim_key]
```

Comment:
Ah, thanks for catching this! I'll delete `projection_dim` for now since I think that could be independent of the text embedding size if that make sense to you as well?